packages/malloy/src/malloy.ts

```diff
@@ -3882,49 +3904,152 @@ export class CSVWriter extends DataWriter {
+      return JSON.stringify(value);
     } else {
-      return `${cell.value}`;
+      return `${value}`;
     }
   }
 
-  async process(data: AsyncIterableIterator<DataRecord>): Promise<void> {
-    let fields;
-    for await (const row of data) {
-      if (fields === undefined) {
-        fields = row.field.allFields;
-        if (this.includeHeader) {
-          for (let fieldIndex = 0; fieldIndex < fields.length; fieldIndex++) {
-            const field = fields[fieldIndex];
-            this.stream.write(this.escape(field.name));
-            if (fieldIndex !== fields.length - 1) {
-              this.stream.write(this.columnSeparator);
-            }
-          }
-          this.stream.write(this.rowSeparator);
+  // Extra weight to be added becase of nested tables inside the cells.
+  private getColWeight(jsonVal: QueryDataRow | QueryData) {
+    let firstVal = jsonVal;
+    if (Array.isArray(jsonVal)) {
+      firstVal = jsonVal[0];
+    }
+    let numKeys = 0;
+    for (const key in firstVal) {
+      numKeys = numKeys + 1;
+      const val = firstVal[key];
+      if (Array.isArray(val)) {
+        const weight = this.getColWeight(val) - 1;
+        numKeys = numKeys + weight;
+      }
+    }
+    return numKeys;
+  }
+
+  // Get header row along with extra empty spaces for nested children.
+  private getHeaderRow(row: QueryDataRow): CellMatrix {
+    let csv = '';
+    let width = 0;
+    for (const key in row) {
+      csv = csv + this.escape(key) + this.columnSeparator;
+      const val = row[key];
+      width++;
+      if (Array.isArray(val)) {
+        const numKeys = this.getColWeight(val) - 1;
+        width = width + numKeys;
+        for (let i = 0; i < numKeys; i++) {
+          csv = csv + this.columnSeparator;
         }
       }
-      for (let fieldIndex = 0; fieldIndex < fields.length; fieldIndex++) {
-        const field = fields[fieldIndex];
-        this.stream.write(this.escape(this.stringify(row.cell(field))));
-        if (fieldIndex !== fields.length - 1) {
-          this.stream.write(this.columnSeparator);
+    }
+    return {rows: [csv], length: 1, width: width};
+  }
+
+  // Merge the child matrices i.e. merge the columns into one bigger matrix i.e. CSV.
+  private mergeMatrices(matrices: CellMatrix[]): CellMatrix {
+    const maxLength = Math.max(...matrices.map(matrix => matrix.length));
+    const matrixWidth = matrices.reduce((sum, matrix) => sum + matrix.width, 0);
+    const csvMatrix: string[] = [];
+    for (let i = 0; i < maxLength; i++) {
+      let csvRow = '';
+      for (const matrix of matrices) {
+        if (i < matrix.length) {
+          csvRow = csvRow + matrix.rows[i];
+          if (!csvRow.endsWith(this.columnSeparator)) {
+            csvRow = csvRow + this.columnSeparator;
+          }
+        } else {
+          // Add empty cells.
+          for (let w = 0; w < matrix.width; w++) {
+            csvRow = csvRow + this.columnSeparator;
```

Comment:
This could be `this.columnSeparator.repeat(matrix.width)` but I would recommend looking into the `.join()` strategy

Reply:
Done.